aea/registries/base.py

```diff
@@ -107,132 +107,107 @@ def teardown(self) -> None:
-        if contract_id not in self._contracts.keys():
-            raise ValueError(
-                "No contract registered with contract id '{}'".format(contract_id)
-            )
-        removed_contract = self._contracts.pop(contract_id)
-        logger.debug("Contract '{}' has been removed.".format(removed_contract.id))
+        Do the actual registration.
 
-    def fetch(self, contract_id: ContractId) -> Optional[Contract]:
+        :param component_id: the component id
+        :param component: the component to register
+        :return: None
         """
-        Fetch the contract by id.
+        self._components_by_type.setdefault(component_id.component_type, {})[
+            component_id.public_id
+        ] = component
+        self._registered_keys.add(component_id)
 
-        :param contract_id: the contract id
-        :return: the contract or None if the contract is not registered
+    def _unregister(self, component_id: ComponentId) -> None:
         """
-        return self._contracts.get(contract_id, None)
-
-    def fetch_all(self) -> List[Contract]:
-        """Fetch all the contracts."""
-        return list(self._contracts.values())
-
-    def setup(self) -> None:
-        """
-        Set up the registry.
+        Do the actual unregistration.
 
+        :param component_id: the component id
         :return: None
         """
-        pass
+        item = self._components_by_type.get(component_id.component_type, {}).pop(
+            component_id.public_id, None
+        )
+        self._registered_keys.discard(component_id)
+        if item is not None:
+            logger.debug("Component '{}' has been removed.".format(item.component_id))
 
-    def teardown(self) -> None:
+    def unregister(self, component_id: ComponentId) -> None:
         """
-        Teardown the registry.
+        Unregister a component.
 
-        :return: None
+        :param component_id: the ComponentId
         """
-        self._contracts = {}
-
-
-class ProtocolRegistry(Registry[PublicId, Protocol]):
-    """This class implements the handlers registry."""
+        if component_id not in self._registered_keys:
+            raise ValueError(
+                "No item registered with contract id '{}'".format(ComponentId)
+            )
+        self._unregister(component_id)
 
-    def __init__(self) -> None:
+    def fetch(self, component_id: ComponentId) -> Optional[Component]:
         """
-        Instantiate the registry.
+        Fetch the component by id.
 
-        :return: None
+        :param component_id: the contract id
+        :return: the component or None if the component is not registered
         """
-        self._protocols = {}  # type: Dict[ProtocolId, Protocol]
+        return self._components_by_type.get(component_id.component_type, {}).get(
+            component_id.public_id, None
+        )
 
-    def register(self, item_id: PublicId, protocol: Protocol) -> None:
+    def fetch_all(self) -> List[Component]:
```

Comment:
Should this not be on a component type basis; i.e. require the component type as arg?

Reply:
Ah I see, exists below